### ISSUE TITLE
project_ownership might have an issue with PHP 8.1

### ISSUE THREAD

**Issue by pbchase** [ctsit/project_ownership#59]
project_ownership might have an issue with PHP 8.1. john.reber reported this in the [REDCap Community](https://redcap.vanderbilt.edu/community/post.php?id=140528):

```
john.reber asked • 4 months, 2 weeks ago (2023-01-11 10:30)
The 'project_ownership' module threw the following exception when calling the hook method 'redcap_every_page_top'
REDCap 13.1.3 · PHP 8.1.14 (Linux/Unix OS) · MariaDB 10.3.13

After upgrading REDCap v12 -> v13.1.3 and PHP from 7.2-> 8.1 we are constantly getting email with the following message:

[This message was automatically generated by REDCap]

The 'project_ownership' module threw the following exception when calling the hook method 'redcap_every_page_top':

Error: Undefined constant "ProjectOwnership\ExternalModule\USERID" in /var/www/html/redcap/modules/project_ownership_v2.0.3/ExternalModule.php:263
Stack trace:
#0 /var/www/html/redcap/modules/project_ownership_v2.0.3/ExternalModule.php(130): ProjectOwnership\ExternalModule\ExternalModule->buildOwnershipFieldset()
#1 /var/www/html/redcap/redcap_v13.1.3/ExternalModules/classes/ExternalModules.php(3254): ProjectOwnership\ExternalModule\ExternalModule->redcap_every_page_top()
#2 /var/www/html/redcap/redcap_v13.1.3/ExternalModules/classes/ExternalModules.php(3403): ExternalModules\ExternalModules::startHook()
#3 /var/www/html/redcap/redcap_v13.1.3/ExternalModules/classes/ExternalModules.php(3436): ExternalModules\ExternalModules::ExternalModules\{closure}()
#4 /var/www/html/redcap/redcap_v13.1.3/Classes/Hooks.php(42): ExternalModules\ExternalModules::callHook()
#5 /var/www/html/redcap/redcap_v13.1.3/Classes/HtmlPage.php(127): Hooks::call()
#6 /var/www/html/redcap/redcap_v13.1.3/Config/init_functions.php(569): HtmlPage->PrintHeader()
#7 /var/www/html/redcap/redcap_v13.1.3/Libraries/PEAR/Auth.php(587): loginFunction()
#8 /var/www/html/redcap/redcap_v13.1.3/Libraries/PEAR/Auth.php(530): Auth->login()
#9 /var/www/html/redcap/redcap_v13.1.3/Classes/Authentication.php(1761): Auth->start()
#10 /var/www/html/redcap/redcap_v13.1.3/Classes/Authentication.php(832): Authentication::checkLogin()
#11 /var/www/html/redcap/redcap_v13.1.3/Classes/Authentication.php(585): Authentication::tableBasedLogin()
#12 /var/www/html/redcap/redcap_v13.1.3/Classes/System.php(739): Authentication::authenticate()
#13 /var/www/html/redcap/redcap_v13.1.3/Config/init_project.php(7): System::initProjectPage()
#14 /var/www/html/redcap/redcap_v13.1.3/ProjectSetup/index.php(4): require_once('...')
#15 {main}

URL: https://redcap.jefferson.edu/redcap_v13.1.3/ProjectSetup/index.php?pid=XXXX
Server: redcap.jefferson.edu (ewebapp02pa.jefferson.edu)
User:
Project ID: XXXX
Module Name: Project Ownership (project_ownership)
Module Author(s): pbc@ufl.edu, tls@ufl.edu, marlycormar@ufl.edu, tbembersimeao@ufl.edu
Run Time: 0 seconds
```

John followed that up with

> [john.reber](https://redcap.vanderbilt.edu/community/profile.php?id=2112) commented • 4 months, 5 days ago (2023-01-25 09:36) 
> We just downgraded PHP from v8.1 to to v7.4, and this and other issues (logging in, DAG issues) we've been having have gone away.

**Comment by pbchase** [comment]
Today I received another, very similar report. This is with the current version of REDCap Project Ownership:

> Hi Philip and team!
> 
> I'm reaching out to see is this being addressed for REDCap 14+ and PHP8+. https://redcap.vanderbilt.edu/community/post.php?id=140528&comment=222849
> We have just migrated to new environment with REDCap 14.1.2, PHP 8.0.30, MySQL 8.0.34. this EM became not functional at all, will get error notification from the PIDs with blank Project Owner, and if you want to add or update, nothing will be saved. We have over 10k projects in total, so we have been getting more and more such error notifications.
> Please kindly advise, thanks!
> 
> Best, 
> Acension REDCap Admins
>  
> 
> ---------- Forwarded message ---------
> From: <redcap@ascension.org>
> Date: Tue, Jan 23, 2024 at 1:14 PM
> Subject: REDCap External Module Hook Exception - project_ownership
> To: <redcap@ascension.org>
> 
> [This message was automatically generated by REDCap]
> 
> The 'project_ownership' module threw the following exception when calling the hook method 'redcap_every_page_top':
> 
> Error: Undefined constant "ProjectOwnership\ExternalModule\USERID" in /home/redcap/public_html/ahnat/modules/project_ownership_v2.4.0/ExternalModule.php:279
> Stack trace:
> #0 /home/redcap/public_html/ahnat/modules/project_ownership_v2.4.0/ExternalModule.php(138): ProjectOwnership\ExternalModule\ExternalModule->buildOwnershipFieldset()
> #1 /home/redcap/public_html/ahnat/redcap_v14.1.2/ExternalModules/classes/ExternalModules.php(3199): ProjectOwnership\ExternalModule\ExternalModule->redcap_every_page_top()
> #2 /home/redcap/public_html/ahnat/redcap_v14.1.2/ExternalModules/classes/ExternalModules.php(3366): ExternalModules\ExternalModules::startHook()
> #3 /home/redcap/public_html/ahnat/redcap_v14.1.2/ExternalModules/classes/ExternalModules.php(3399): ExternalModules\ExternalModules::ExternalModules\{closure}()
> #4 /home/redcap/public_html/ahnat/redcap_v14.1.2/Classes/Hooks.php(41): ExternalModules\ExternalModules::callHook()
> #5 /home/redcap/public_html/ahnat/redcap_v14.1.2/Classes/HtmlPage.php(150): Hooks::call()
> #6 /home/redcap/public_html/ahnat/redcap_v14.1.2/Config/init_functions.php(656): HtmlPage->PrintHeader()
> #7 /home/redcap/public_html/ahnat/redcap_v14.1.2/Libraries/PEAR/Auth.php(587): loginFunction()
> #8 /home/redcap/public_html/ahnat/redcap_v14.1.2/Libraries/PEAR/Auth.php(530): Auth->login()
> #9 /home/redcap/public_html/ahnat/redcap_v14.1.2/Classes/Authentication.php(1859): Auth->start()
> #10 /home/redcap/public_html/ahnat/redcap_v14.1.2/Classes/Authentication.php(894): Authentication::checkLogin()
> #11 /home/redcap/public_html/ahnat/redcap_v14.1.2/Classes/Authentication.php(647): Authentication::tableBasedLogin()
> #12 /home/redcap/public_html/ahnat/redcap_v14.1.2/Classes/System.php(936): Authentication::authenticate()
> #13 /home/redcap/public_html/ahnat/redcap_v14.1.2/Config/init_project.php(7): System::initProjectPage()
> #14 /home/redcap/public_html/ahnat/redcap_v14.1.2/ProjectSetup/index.php(4): require_once('...')
> #15 {main}
> 
> URL: https://redcap.ascension.org/ahnat/redcap_v14.1.2/ProjectSetup/index.php?pid=4650
> Server: redcap.ascension.org (andc1lweb0004)
> User: 
> Project ID: 4650
> Module Name: Project Ownership (project_ownership)
> Module Author(s): CTSIT-REDCAP-MODULE-SUPPO@LISTS.UFL.EDU
> Run Time: 0 seconds

**Comment by ChemiKyle** [comment]
I can reproduce this by attempting to navigate to any project setup page from a private browsing setting (i.e. while not logged in).

**Comment by pbchase** [comment]
Addressed by PR #66

**Comment by pbchase** [comment]
I submitted release 2.4.1 to the REDCap Repo on 2024-02-06
